File: client/lib/post.ts
import Comment from "./comment";

export default class Post {
  id: number;
  author: string;
  author_id: number;
  author_pname: string;
  content: string;
  date_created: Date;

  avatar?: string;
  date_updated?: Date;
  comments?: Comment[];
  topics?: string[];
  mentions?: string[];

  constructor(
    id: number,
    author: string,
    author_pname: string,
    author_id: number,
    content: string,
    date_created: Date
  ) {
    this.id = id;
    this.author = author;
    this.author_pname = author_pname;
    this.author_id = author_id;
    this.content = content;
    this.date_created = new Date(date_created);
  }

  extractTopics(): Array<string> {
    let topics: Array<string> = [];

    this.content.split(" ").forEach((value) => {
      if (value[0] == "#") topics.push(value);
    });

    return topics;
  }

  extractMentions(): Array<string> {
    let mentions: Array<string> = [];

    this.content.split(" ").forEach((value) => {
      if (value[0] == "@") mentions.push(value);
    });

    return mentions;
  }
}

File: server/internal/models/post.go
package models

import (
	"database/sql"
	"encoding/json"
)

type Post struct {
	ID          int64          `json:"id" db:"post_id"`
	Author      string         `json:"author" db:"username"`
	AuthorPName string         `json:"author_pname" db:"profile_name"`
	Avatar      string         `json:"avatar" db:"avatar"`
	Author_ID   int64          `json:"author_id" db:"user_id"`
	Content     string         `json:"content" db:"content"`
	DateCreated string         `json:"date_created" db:"date_created"`
	DateUpdated JsonNullString `json:"date_updated" db:"date_updated"`
}

type Topic struct {
	Id    int64  `json:"topic_id" db:"topic_id"`
	Topic string `json:"topic_name" db:"topic_name"`
}

type JsonNullString struct {
	sql.NullString
}

func (v JsonNullString) MarshalJSON() ([]byte, error) {
	if v.Valid {
		return json.Marshal(v.String)
	} else {
		return json.Marshal(nil)
	}
}

func (v *JsonNullString) UnmarshalJSON(data []byte) error {
	var x *string
	if err := json.Unmarshal(data, &x); err != nil {
		return err
	}
	if x != nil {
		v.Valid = true
		v.String = *x
	} else {
		v.Valid = false
	}
	return nil
}
